lib/compiler/aro_translate_c.zig

```diff
@@ -215,80 +218,292 @@ fn prepopulateGlobalNameTable(c: *Context) !void {
+            try transEnumDecl(c, scope, decl, fields[0..field_count]);
+        },
+        .enum_decl => {
+            const fields = c.tree.data[data.range.start..data.range.end];
+            try transEnumDecl(c, scope, decl, fields);
+        },
 
-            .fn_proto,
-            .static_fn_proto,
-            .inline_fn_proto,
-            .inline_static_fn_proto,
-            .fn_def,
-            .static_fn_def,
-            .inline_fn_def,
-            .inline_static_fn_def,
-            => {
-                try transFnDecl(c, node);
-            },
+        .enum_field_decl,
+        .record_field_decl,
+        .indirect_record_field_decl,
+        .struct_forward_decl,
+        .union_forward_decl,
+        .enum_forward_decl,
+        => return,
+
+        .fn_proto,
+        .static_fn_proto,
+        .inline_fn_proto,
+        .inline_static_fn_proto,
+        .fn_def,
+        .static_fn_def,
+        .inline_fn_def,
+        .inline_static_fn_def,
+        => {
+            try transFnDecl(c, decl);
+        },
 
-            .@"var",
-            .static_var,
-            .threadlocal_var,
-            .threadlocal_static_var,
-            .extern_var,
-            .threadlocal_extern_var,
-            => {
-                try transVarDecl(c, node, null);
-            },
-            else => unreachable,
-        }
+        .@"var",
+        .extern_var,
+        .static_var,
+        .threadlocal_var,
+        .threadlocal_extern_var,
+        .threadlocal_static_var,
+        => {
+            try transVarDecl(c, decl, null);
+        },
+        else => unreachable,
     }
 }
 
 fn transTypeDef(_: *Context, _: *Scope, _: NodeIndex) Error!void {
     @panic("TODO");
 }
-fn transRecordDecl(_: *Context, _: *Scope, _: NodeIndex) Error!void {
-    @panic("TODO");
+
+fn mangleWeakGlobalName(c: *Context, want_name: []const u8) ![]const u8 {
+    var cur_name = want_name;
+
+    if (!c.weak_global_names.contains(want_name)) {
+        // This type wasn't noticed by the name detection pass, so nothing has been treating this as
+        // a weak global name. We must mangle it to avoid conflicts with locals.
+        cur_name = try std.fmt.allocPrint(c.arena, "{s}_{d}", .{ want_name, c.getMangle() });
+    }
+
+    while (c.global_names.contains(cur_name)) {
+        cur_name = try std.fmt.allocPrint(c.arena, "{s}_{d}", .{ want_name, c.getMangle() });
+    }
+    return cur_name;
+}
+
+fn transRecordDecl(c: *Context, scope: *Scope, record_node: NodeIndex, field_nodes: []const NodeIndex) Error!void {
+    const node_types = c.tree.nodes.items(.ty);
+    const raw_record_ty = node_types[@intFromEnum(record_node)];
+    const record_decl = raw_record_ty.getRecord().?;
+    if (c.decl_table.get(@intFromPtr(record_decl))) |_|
+        return; // Avoid processing this decl twice
+    const toplevel = scope.id == .root;
+    const bs: *Scope.Block = if (!toplevel) try scope.findBlockScope(c) else undefined;
+
+    var is_union = false;
```

Comment:
```suggestion
    const container_kind: ZigTag = if (raw_record_ty.is(.@"union")) .@"union" else .@"struct";
```

Then `container_kind_name` usage can be replaced with `@tagName(container_kind)`; the record payload `{ .tag = ([2]ZigTag{ .@"struct", .@"union" })[@intFromBool(is_union)] }` can be replaced with `.{ .tag = container_kind }`, and the `default_value` node condition can check `container_kind != .@"union"`